File: packages/typescript/sdk/core/src/diff.ts
import { escapeJsonPointerSegment } from "./paths";
import type { PatchOp, SlopNode } from "./types";

/**
 * Recursively diff two SLOP trees and produce JSON Patch ops.
 * Paths use node IDs in children segments (not array indices).
 */
export function diffNodes(oldNode: SlopNode, newNode: SlopNode, basePath: string = ""): PatchOp[] {
  const ops: PatchOp[] = [];

  // Diff properties key-by-key
  const oldProps = oldNode.properties ?? {};
  const newProps = newNode.properties ?? {};
  const allKeys = new Set([...Object.keys(oldProps), ...Object.keys(newProps)]);
  for (const key of allKeys) {
    const oldVal = oldProps[key];
    const newVal = newProps[key];
    const escapedKey = escapeJsonPointerSegment(key);
    if (oldVal === undefined && newVal !== undefined) {
      ops.push({ op: "add", path: `${basePath}/properties/${escapedKey}`, value: newVal });
    } else if (oldVal !== undefined && newVal === undefined) {
      ops.push({ op: "remove", path: `${basePath}/properties/${escapedKey}` });
    } else if (JSON.stringify(oldVal) !== JSON.stringify(newVal)) {
      ops.push({ op: "replace", path: `${basePath}/properties/${escapedKey}`, value: newVal });
    }
  }

  // Diff affordances (replace entire array if changed)
  if (JSON.stringify(oldNode.affordances) !== JSON.stringify(newNode.affordances)) {
    if (newNode.affordances) {
      ops.push({
        op: oldNode.affordances ? "replace" : "add",
        path: `${basePath}/affordances`,
        value: newNode.affordances,
      });
    } else if (oldNode.affordances) {
      ops.push({ op: "remove", path: `${basePath}/affordances` });
    }
  }

  // Diff meta (replace entire object if changed)
  if (JSON.stringify(oldNode.meta) !== JSON.stringify(newNode.meta)) {
    if (newNode.meta) {
      ops.push({
        op: oldNode.meta ? "replace" : "add",
        path: `${basePath}/meta`,
        value: newNode.meta,
      });
    } else if (oldNode.meta) {
      ops.push({ op: "remove", path: `${basePath}/meta` });
    }
  }

  // Diff content_ref (replace entire object if changed)
  if (JSON.stringify(oldNode.content_ref) !== JSON.stringify(newNode.content_ref)) {
    if (newNode.content_ref) {
      ops.push({
        op: oldNode.content_ref ? "replace" : "add",
        path: `${basePath}/content_ref`,
        value: newNode.content_ref,
      });
    } else if (oldNode.content_ref) {
      ops.push({ op: "remove", path: `${basePath}/content_ref` });
    }
  }

  // Diff children — simulate apply order so indices stay accurate.
  const oldChildren = oldNode.children ?? [];
  const newChildren = newNode.children ?? [];
  const oldMap = new Map(oldChildren.map((c) => [c.id, c]));
  const newMap = new Map(newChildren.map((c) => [c.id, c]));

  // 1. Remove children no longer present.
  const working: string[] = [];
  for (const child of oldChildren) {
    if (!newMap.has(child.id)) {
      ops.push({ op: "remove", path: `${basePath}/${child.id}` });
    } else {
      working.push(child.id);
    }
  }

  // 2. Insert new children at their target positions.
  for (let i = 0; i < newChildren.length; i++) {
    const child = newChildren[i];
    if (!oldMap.has(child.id)) {
      ops.push({ op: "add", path: `${basePath}/${child.id}`, value: child, index: i });
      working.splice(i, 0, child.id);
    }
  }

  // 3. Emit move ops until the working order matches the target order.
  for (let i = 0; i < newChildren.length; i++) {
    const targetId = newChildren[i].id;
    if (working[i] === targetId) continue;
    const currentIdx = working.indexOf(targetId, i);
    if (currentIdx === -1) continue;
    ops.push({ op: "move", path: `${basePath}/${targetId}`, index: i });
    working.splice(currentIdx, 1);
    working.splice(i, 0, targetId);
  }

  // 4. Recurse into kept children.
  for (const child of newChildren) {
    const oldChild = oldMap.get(child.id);
    if (oldChild) {
      ops.push(...diffNodes(oldChild, child, `${basePath}/${child.id}`));
    }
  }

  return ops;
}

File: packages/swift/slop-ai/Sources/SlopAI/Diff.swift
import Foundation

/// Recursively diff two SLOP trees and produce JSON Patch-like operations.
/// Child paths use node IDs rather than array indices, matching the TS SDK.
public func diffNodes(_ oldNode: SlopNode, _ newNode: SlopNode, basePath: String = "") -> [PatchOp] {
  var ops: [PatchOp] = []

  let oldProperties = oldNode.properties ?? [:]
  let newProperties = newNode.properties ?? [:]
  for key in Set(oldProperties.keys).union(newProperties.keys).sorted() {
    let escapedKey = escapeJSONPointerSegment(key)
    let path = "\(basePath)/properties/\(escapedKey)"
    switch (oldProperties[key], newProperties[key]) {
    case (nil, .some(let value)):
      ops.append(PatchOp(op: .add, path: path, value: value))
    case (.some, nil):
      ops.append(PatchOp(op: .remove, path: path))
    case (.some(let oldValue), .some(let newValue)) where oldValue != newValue:
      ops.append(PatchOp(op: .replace, path: path, value: newValue))
    default:
      break
    }
  }

  if oldNode.affordances != newNode.affordances {
    if let affordances = newNode.affordances {
      ops.append(PatchOp(op: oldNode.affordances == nil ? .add : .replace, path: "\(basePath)/affordances", value: wireJSON(affordances)))
    } else if oldNode.affordances != nil {
      ops.append(PatchOp(op: .remove, path: "\(basePath)/affordances"))
    }
  }

  if oldNode.meta != newNode.meta {
    if let meta = newNode.meta {
      ops.append(PatchOp(op: oldNode.meta == nil ? .add : .replace, path: "\(basePath)/meta", value: wireJSON(meta)))
    } else if oldNode.meta != nil {
      ops.append(PatchOp(op: .remove, path: "\(basePath)/meta"))
    }
  }

  if oldNode.contentRef != newNode.contentRef {
    if let contentRef = newNode.contentRef {
      ops.append(PatchOp(op: oldNode.contentRef == nil ? .add : .replace, path: "\(basePath)/content_ref", value: wireJSON(contentRef)))
    } else if oldNode.contentRef != nil {
      ops.append(PatchOp(op: .remove, path: "\(basePath)/content_ref"))
    }
  }

  let oldChildren = oldNode.children ?? []
  let newChildren = newNode.children ?? []
  let oldMap = oldChildren.reduce(into: [String: SlopNode]()) { map, child in
    map[child.id] = child
  }
  let newMap = newChildren.reduce(into: [String: SlopNode]()) { map, child in
    map[child.id] = child
  }

  var working: [String] = []
  for child in oldChildren {
    if newMap[child.id] == nil {
      ops.append(PatchOp(op: .remove, path: "\(basePath)/\(child.id)"))
    } else {
      working.append(child.id)
    }
  }

  for (index, child) in newChildren.enumerated() where oldMap[child.id] == nil {
    ops.append(PatchOp(op: .add, path: "\(basePath)/\(child.id)", value: wireJSON(child), index: index))
    working.insert(child.id, at: min(index, working.count))
  }

  for (index, child) in newChildren.enumerated() where working.indices.contains(index) && working[index] != child.id {
    guard let currentIndex = working[index...].firstIndex(of: child.id) else { continue }
    ops.append(PatchOp(op: .move, path: "\(basePath)/\(child.id)", index: index))
    working.remove(at: currentIndex)
    working.insert(child.id, at: index)
  }

  for child in newChildren {
    if let oldChild = oldMap[child.id] {
      ops.append(contentsOf: diffNodes(oldChild, child, basePath: "\(basePath)/\(child.id)"))
    }
  }

  return ops
}
